README.md

```diff
@@ -0,0 +1,396 @@
+
+Plugins always "plug in" to some "manager" component (typically named `foomgr`), and should depend on that manager and register themselves with it at startup.
+Then, it is up to apps to depend on the necessary plugins.
+
+## Programming Errors
+
+Programming errors, such as calling a method at an inappropriate time, should be handled with `panic(..)` instead of errors.
+If an error is returned, it will likely be logged and may not be seen.
+A panic, on the other hand, is very noticeable and carries a stack trace that can help the programmer figure out what they've missed.
+Try to arrange for such panics to happen consistently, so that such programming errors are quick to find.
+
+# Future Plans
+
+## Component Linting
+
+With good detection of components (already used to generate COMPONENTS.md and CODEOWNERS), we can check that the guidelines are followed.
+For example, this check could easily verify that components are not nested, and that every component has a `Component` type and `Module` value.
+
+## Config Reducers
+
+We can use a concept similar to that defined by Redux to simplify the DD configuration used by each component that needs it.
+This would involve a "reducer" that extracts data from Viper (or whatever we switch to) and places it in a component-specific struct.
+Using struct tags and a utility function would allow for a very regular, greppable arrangement of configuration parameters with very little per-component boilerplate.
+This will also ease testing of components: tests can simply provide a filled-in configuration struct, instead of manually setting configuration parameters.
+
+# Open Questions
+
+## Subprocesses
+
+We want to support running some "things" (we should have a term for this!) as subprocesses, as is currently done for trace-agent, system-probe, process-agent, and security-agent.
+Should these be different binaries (as they are now), or the same binary with different arguments?
+
+## Representing Phases with Types
+
+Olivier Gaça suggests that we could avoid warnings in comments to only call certain methods before/after startup by using distinct types to represent the component before and after startup.
+
+## ModuleParams Defaults
+
+Currently, the guide recommends that `ModuleParams` be optional and that its zero value be meaningful.
+But this may result in a lot of `if params.Xyz == "" { params.Xyz = "real default" }`, which could be annoying and ugly.
+If component users were encouraged to use `fx.Replace(foo.ModuleParams{ Xyz: "specific value" })` to specify parameters, then the component's module could use `fx.Supply(ModuleParams{ Xyz: "real default", ... }` to supply nonzero defaults.
+
+## Enabling / Disabling Component Bundles
+
+Consider configs like `logs_enabled` or `apm_config.enabled` -- when these are false, it means that several components that are already instantiated should not actually start.
+One options here is to thread a bunch of `Enable` methods through these components.
```

Comment:
Thanks -- those are both good ideas.  I'll play around some more and see if I can make the pieces fit together.

Reply:
Depending on the number of subcomponents having an `Enable` method may be a good choice too. I suggested alternative approaches because the project you did is a wonderful playground to try ideas and see the result.

Reply:
I guess it somewhat overlaps with the discussion about what should be a component or not. The root question is how does the app entrypoint knows which component it should register.

Currently it has two inputs to know what to do: exec name and config. I agree with Olivier that `configuration` will need a special treatment so that it can be used to instantiate components.
Maybe we need a common code that helps with that.

Reply:
There are actually two steps here:
```
fx.New(foo.Module, bar.Module, ..)
```
doesn't _instantiate_ anything -- it just defines some providers.  It's when those providers are required that the instantiation happens.  So if by "register" you mean "pass an fx.Provides(..) to fx.New", the answer to "which component it should register" is "all of the components in this binary", which is a value known at build time.

The question of what to _require_ (via an fx.Invoke) is harder.  Olivier's first suggestion would result in never instantiating the logs-agent component (and all components it requires) if `logs_config.enabled` is false, but the cost is that configuration must be loaded before the App is created -- something I'd like to avoid (global configuration has bitten us before!).

The other two options are to always instantiate the logs-agent component, but instruct it to do nothing when it's not enabled.  We can either check that in the app, or put the check within the logs-agent.

Reply:
I am not sure I agree about:
```
the answer to "which component it should register" is "all of the components in this binary", which is a value known at build time.
```

The first thing is how does the `cmd` knows all existing components, except for developers to add them to all `cmd` whenever a new component is created?

If we consider that we should add only top-level bundles in `cmd`, it means that the `bundle` will choose to provide for all second level components (e.g. components that a `bundle` depends on but not registered in `cmd`)
How does it work if two bundles require different implementations of the same component?

If we consider that everything should be registered in `cmd`, it means we'll indeed have tons of components linked in all `cmd`.

I agree that the question of what needs to be `Invoke` is also interesting, but I'm not sure we solved the `Provide` one entirely.

Reply:
> how does the cmd knows all existing components? .. If .. we should add only top-level bundles in cmd,

Yes, that's my intended answer.

> How does it work if two bundles require different implementations of the same component?

I am not anticipating different implementations of the same component, partially for this reason.  So, if that's for example remote tagger and local tagger, those two would have different interfaces (remotetagger.Component and localtagger.Component).  Maybe we have a wrapper that can use either one and supports fallback.  So `comp/tagger` may supply all three, and it's up to other components which one to require.

The central idea is that the set of types that are provided, and their implementations, are the same on every invocation -- regardless of configuration.  Anything that needs to be dynamic needs to happen at a later phase.

There are three points where we can decide whether or not to "use" a component:
 * Provide (`fx.Provide(constructor)`)
 * Instantiate (`fx.Invoke` or require from a constructor)
 * Start (related to my other comments about enabling components)

The first has a cost in terms of binary size, but is otherwise very cheap -- providing a type that is never instantiated has no runtime penalty.  So that's why my answer is "all of the components in this binary", and why I think it's OK (and simple) to make this a static value.  [The current implementation is actually a little more complex, in that each subcommand (`agent run`, `agent status`, etc.) is defining its own fx.App with a different set of provided types, but I think that's actually easier to reason about].

The second is basically free -- it allocates a small struct, perhaps, but no more.

The third has a cost if the component actually starts, and a cost in complexity to communicate in which circumstances it should start.

Looked at another way: _Provide ⊃ Instantiate ⊃ Start_.  _Provide_ is defined at build time.  The size of _Provide - Instantiate_ is relevant only in terms of binary size, and the size of _Instantiate - Start_ is trivial.